rfd/0144-client-tools-auto-updates.md

```diff
@@ -0,0 +1,94 @@
+---
+authors: Bernard Kim (bernard@goteleport.com)
+state: draft
+---
+
+# RFD 0144 - Client tools auto update
+
+## Required Approvers
+
+* Engineering: @r0mant && @jimbishopp
+* Product: @klizhentas || @xinding33
+* Security: @jentfoo
+
+## What
+
+Support for Teleport client tools (both tsh and tctl) to automatically stay up to date with the latest cloud-stable version of Teleport.
+
+## Why
+
+One of our goals for Teleport Cloud is to increase the cadence at which clusters can be updated.
+This requires that Teleport Agents and Teleport client tools be up to date with the latest compatible version of the cluster.
+
+Manually updating Teleport Agents and client tools can be a tedious task, so customers can end up delaying their updates.
+Simplifying this process will help customers update and stay up to date with their Teleport version.
+
+Development for Automatic Updates of Teleport Agents has already been completed, so the success criteria for this RFD only includes developing a simpler way to keep Teleport client tools up to date.
+
+## Details
+
+Teleport will now store a single cached version of the client tools in the user's $HOME directory. tsh/tctl will execute this cached version when communicating with the Teleport Cloud cluster.
+
+Cloud will host an endpoint for Teleport client tools version discovery. tsh/tctl will check this endpoint to identify the current cloud-stable version of Teleport. If the cached version is behind or ahead of the current cloud-stable version, the cache will be updated to match the current cloud-stable version. This will ensure Teleport client tools are always executed using the latest compatible version of Teleport when communicating with a Teleport Cloud cluster.
+
+### Detecting Cloud instances
+
+For now this feature will be disabled for self-hosted users and will only be supported for Cloud users.
+
+For tsh/tctl to identify if it is communicating with a Cloud instance of Teleport the `../webapi/ping` endpoint will now include a `cloud` boolean flag in it's ping response.
```

Comment:
Will this require the minor version to be disclosed by `ping`?

Having this unauthenticated version disclosure is not ideal since it can help attackers target attacks against unpatched customers: https://github.com/gravitational/teleport-private/issues/161

We have discussed removing the minor version from the `ping` endpoint, but there has not been any action to it yet.

Ideally the version would only be disclosed to an authenticated user, what do you think about doing that authentication first?

Reply:
I don't think the proposed design will require the minor version be disclosed by ping. The client tools will be querying the endpoint described to get the latest stable version of teleport.